### ISSUE TITLE
poor WASD, arrow keys performance on Windows, Firefox

### ISSUE THREAD

**Issue by lmulhall-phet** [phetsims/balloons-and-static-electricity#402]
<b>Test Device</b>

Hanson

<b>Operating System</b>

Windows 10

<b>Browser</b>

Firefox 60.0

<b>Problem Description</b>

Related to phetsims/QA/issues/112. Using WASD or the arrow keys to move a balloon while NVDA (EDIT: or JAWS) is running is extremely laggy. (This might be a machine-specific issue.)

<b>Steps to Reproduce</b>

1. Have NVDA (EDIT: or JAWS) running.
2. Move balloon using WASD or the arrow keys.

<b>Visuals</b>

![base](https://user-images.githubusercontent.com/31285590/40078603-4e46bd56-5842-11e8-84da-7fe7617f19c0.gif)

Note: I had a screen recorder and NVDA running when I recorded this. If I'm only running NVDA, the performance isn't as bad, but it's still pretty bad.

<details>
<summary><b>Troubleshooting Information</b></summary>

Name: ‪Balloons and Static Electricity‬
URL: https://bayes.colorado.edu/dev/html/balloons-and-static-electricity/1.4.0-dev.21/phet/balloons-and-static-electricity_all_phet_debug.html?accessibility
Version: 1.4.0-dev.21 2018-04-04 18:25:56 UTC
Flags: pixelRatioScaling
User Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:60.0) Gecko/20100101 Firefox/60.0
Language: en-US
Window: 1280x606
Pixel Ratio: 2.5/1
WebGL: WebGL 1.0
GLSL: WebGL GLSL ES 1.0
Vendor: Mozilla (Mozilla)
Vertex: attribs: 16 varying: 30 uniform: 4096
Texture: size: 16384 imageUnits: 16 (vertex: 16, combined: 32)
Max viewport: 32767x32767
OES_texture_float: true
Dependencies JSON: {"assert":{"sha":"928741cf","branch":"master"},"axon":{"sha":"ceb3cc00","branch":"master"},"balloons-and-static-electricity":{"sha":"f4f3cc64","branch":"master"},"brand":{"sha":"89d28f63","branch":"master"},"chipper":{"sha":"bc7c5d1f","branch":"master"},"dot":{"sha":"e5489d18","branch":"master"},"joist":{"sha":"2cf256a9","branch":"master"},"kite":{"sha":"b6071478","branch":"master"},"phet-core":{"sha":"d8804eee","branch":"master"},"phet-io":{"sha":"4bbddfaa","branch":"master"},"phet-io-website":{"sha":"b6c2b0f5","branch":"master"},"phet-io-wrapper-classroom-activity":{"sha":"1ecd81a1","branch":"master"},"phet-io-wrapper-lab-book":{"sha":"b1c5562c","branch":"master"},"phet-io-wrappers":{"sha":"025cc23d","branch":"master"},"phetcommon":{"sha":"de004c6e","branch":"master"},"query-string-machine":{"sha":"c379f1db","branch":"master"},"scenery":{"sha":"4a4afee2","branch":"master"},"scenery-phet":{"sha":"e68ebc03","branch":"master"},"sherpa":{"sha":"bb14a03a","branch":"master"},"sun":{"sha":"8a187c9f","branch":"master"},"tandem":{"sha":"f8147666","branch":"master"},"vibe":{"sha":"2dc20be7","branch":"master"}}

</details>

**Comment by jessegreenberg** [comment]
Thanks @lmulhall-phet, that is very bad. Firefox performance is much better on my machine, but I have been testing on FF ESR 52. Firefox 60 has been known to significantly reduce performance for accessibility. @lmulhall-phet I am sorry I didn't mention this. I would like to 

- Try FF 60 on another device to see if it is faster.
- Try FF ESR on Hanson to see if that significantly improves performance. (I understand this may not be possible if we don't want to install a different version of FF on Hanson.)

Depending on performance this will give us data about whether we want to spend time optimizing or wait until FF improves Quantum for a11y.

**Comment by jessegreenberg** [comment]
Testing on FF 60 with NVDA using @phet-steele's laptop.

I don't notice bad framerate, but FF 60 with NVDA does have a problem where the WASD and Arrow keys behave as though they remain pressed down for longer than they are. This could also be performance related.

**Comment by jessegreenberg** [comment]
@lmulhall-phet is going to try with FF ESR 52 on Hanson to see if that improves performance. Thanks!

**Comment by lmulhall-phet** [comment]
@jessegreenberg the WASD / arrow key performance with NVDA is markedly better with FF ESR 52. Same for JAWS. I'm not experiencing the phenomenon where the balloon would continue to move as if you were still pressing one of the keys to move it.

**Comment by jessegreenberg** [comment]
Thanks @lmulhall-phet, good news. In that case, we should discuss with @emily-phet and @terracoda about how much time PhET should spend improving improving performance due to FF Quantum.

On one hand, FF is up to version 60, and has  been trying to bring people to latest version since Jan of 2018 (https://www.marcozehe.de/2018/01/16/nvda-firefox-58-team-regaining-strength/). As of Jan of 2018, Mozilla says:

> Statistics quickly showed that about two thirds of the NVDA user base stayed with [FF Quantum] despite this.

But on the other hand, they clearly still have a lot of work to do. The latest version of JAWS *still* complains that FF 60 is incompatible and  FF ESR 52 should be used. So I would expect them to continue improving performance on their side.

So I would recommend not devoting time to platform specific performance improvements while latest FF a11y is too unstable to even support basic JAWS use.

@terracoda @emily-phet, thoughts?

**Comment by terracoda** [comment]
@jessegreenberg, I realize we do not officially support JAWS with Chrome, IE or Edge, but would it be prudent just to test a little in each of the browsers since performance in FF has decreased significantly? 

I am not really sure how to deal with this AT support flux?

**Comment by jessegreenberg** [comment]
I don't think so @terracoda, the benefit of working on a small set of supported platforms is that we don't have to spend resources testing and fixing bugs on platforms that are not in the list. I think we should focus on full support or none for a browser.

**Comment by terracoda** [comment]
Ok, understood.

**Comment by jessegreenberg** [comment]
It sounds like we will likely be targeting FF 60. Reassigning to myself.

**Comment by jessegreenberg** [comment]
When I test with NVDA and FF 60, the it acts as though I am holding  down the WASD/arrow keys for 0.5-1.0 seconds longer than I am (sometimes).

I am not seeing the substantial loss of framerate shown in the issue ticket.

**Comment by lmulhall-phet** [comment]
@jessegreenberg I am seeing the same behavior. Do you think we ought to close this issue and open a new one for the 0.5s longer arrow key press issue? Or do you think they are related somehow?

**Comment by jessegreenberg** [comment]
Thanks @lmulhall-phet, it sounds like this issue can be closed, Ill make a new one now to track the delay. If it is also only 0.5 seconds on Hanson it might not be worth spending too much time on.

**Comment by terracoda** [comment]
@lmulhall-phet and @jessegreenberg, I am fine with closing this issue.

**Comment by jessegreenberg** [comment]
Sounds good @terracoda, closing. Ill create a new one to track the 0.5 second delay.
